In [1]:
import pandas as pd
train = pd.read_csv("Train_Data.csv") #read train csv

In [2]:
train

headline  is_sarcastic
0      supreme court votes 7-2 to legalize all worldl...             1
1      hungover man horrified to learn he made dozens...             1
2      emily's list founder: women are the 'problem s...             0
3          send your kids back to school with confidence             0
4              watch: experts talk pesticides and health             0
...                                                  ...           ...
44257   greece seeks to reassure europe as tensions rise             0
44258  vatican says transgender man cannot become a g...             0
44259  protesters ejected from donald trump rally aft...             0
44260        italian recipes that are oldies but goodies             0
44261  area loser blissfully unaffected by whims of s...             1

[44262 rows x 2 columns]

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44262 entries, 0 to 44261
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   headline      44262 non-null  object
 1   is_sarcastic  44262 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 691.7+ KB


In [4]:
train.isnull().values.any() #check any null value or not in train csv

False

In [5]:
train['headline'].iloc[1]

'hungover man horrified to learn he made dozens of plans last night'

# Text Preprocessing

In [6]:
train['headline']= train['headline'].str.replace("\r", " ") #replacing \r

In [7]:
train['headline']= train['headline'].str.replace("\n", " ") #replacing \n

In [8]:
train['headline']= train['headline'].str.replace("    ", " ")  #replacing double spaces

In [9]:
train['headline']= train['headline'].str.replace('"', '') #replacing ''

In [10]:
train['headline'].iloc[1]

'hungover man horrified to learn he made dozens of plans last night'

In [11]:
train['headline']= train['headline'].str.lower() #lowering casing the text 

In [12]:
#cleaning punctuation signs 
punctuation_signs = list("?:!.,;")
train['headline'] = train['headline']

for i in punctuation_signs:
    train['headline'] = train['headline'].str.replace(i, '')

C:\Users\nyach\AppData\Local\Temp\ipykernel_10256\800866778.py:6: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  train['headline'] = train['headline'].str.replace(i, '')


In [13]:
train['headline'].iloc[0]

'supreme court votes 7-2 to legalize all worldly vices'

In [14]:
#cleaning possessive pronouns
train['headline'] = train['headline'].str.replace("'s", "")

In [15]:
# Downloading the stop words list
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\nyach\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [16]:
from nltk.corpus import stopwords

# Loading the stop words in english
stop_words = list(stopwords.words('english'))

In [17]:
stop_words

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [18]:
#removing stop words
for i in stop_words:
    stopword = r"\b" + i + r"\b"
    train['headline'] = train['headline'].str.replace(stopword, '')

C:\Users\nyach\AppData\Local\Temp\ipykernel_10256\3243000753.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  train['headline'] = train['headline'].str.replace(stopword, '')


In [19]:
train['headline'].iloc[0]

'supreme court votes 7-2  legalize  worldly vices'

In [20]:
train['headline'].iloc[1]

'hungover man horrified  learn  made dozens  plans last night'

In [21]:
#after text preprocessing
train 

headline  is_sarcastic
0       supreme court votes 7-2  legalize  worldly vices             1
1      hungover man horrified  learn  made dozens  pl...             1
2      emily list founder women   'problem solvers'  ...             0
3                    send  kids back  school  confidence             0
4                  watch experts talk pesticides  health             0
...                                                  ...           ...
44257       greece seeks  reassure europe  tensions rise             0
44258  vatican says transgender man cannot become  go...             0
44259  protesters ejected  donald trump rally  holdin...             0
44260                  italian recipes   oldies  goodies             0
44261  area loser blissfully unaffected  whims  stock...             1

[44262 rows x 2 columns]

# Split Train_data.csv 

In [24]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(train['headline'],train['is_sarcastic'],
                                                    test_size=0.3, 
                                                    random_state=42)

In [25]:
X_train

10080    van jones tough-guy 'trumpzilla'  become 'pres...
32437    us–cuba relations end  obama hit  foul ball  e...
34418                     7 phrases   never say   argument
8688              teen anxious  cigarette addiction  kick 
17114    god irritated guests   understand  time  leave...
                               ...                        
6265      consumer financial protection bureau  governm...
11284                    never  -male panels ubs exec says
38158    ' sleep  target chic new modern home collectio...
860        6 incredible photos  show  world  need  protect
15795     seeing  fifth-grader get bullied  group  boys...
Name: headline, Length: 30983, dtype: object

In [26]:
X_test

12782                       north dakota  heard   48 hours
42915    report  going  take way    inconceivable act  ...
33043    states' rights rancher ryan bundy  run  nevada...
1121     watching thousands march   honor unlocks deepe...
38782           debate  two   unthinkable  united  country
                               ...                        
29585    google combats holocaust-denying search result...
10714    chrissy teigen hot take  ice cream trends coul...
41705                fell  love carrying another man child
31786    mom hilarious story   morning shows  hectic li...
3382     report underpaid migrant laborers working 18 h...
Name: headline, Length: 13279, dtype: object

In [27]:
y_train

10080    0
32437    1
34418    0
8688     1
17114    1
        ..
6265     0
11284    0
38158    0
860      0
15795    0
Name: is_sarcastic, Length: 30983, dtype: int64

In [28]:
y_test

12782    1
42915    1
33043    0
1121     1
38782    0
        ..
29585    0
10714    0
41705    0
31786    0
3382     1
Name: is_sarcastic, Length: 13279, dtype: int64

In [36]:
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((30983,), (13279,), (30983,), (13279,))

# Extract TF-IDF Features

In [35]:
from sklearn.feature_extraction.text import TfidfVectorizer

#parameter
ngram_range = (1,2) #both unigrams and bigrams 
min_df = 1
max_df = 100
max_features = 15000
tfidf = TfidfVectorizer(encoding='utf-8',
                        ngram_range=ngram_range,
                        stop_words=None,
                        lowercase=False,
                        max_df=max_df,
                        min_df=min_df,
                        max_features=max_features,
                        norm='l2',
                        sublinear_tf=True)
                        
features_train = tfidf.fit_transform(X_train).toarray()
labels_train = y_train
print(features_train.shape)

features_test = tfidf.transform(X_test).toarray()
labels_test = y_test
print(features_test.shape)

(30983, 15000)
(13279, 15000)


# Using various sklearn models to compare accuracy among them

In [39]:
##naive bias
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import CountVectorizer 

nb = Pipeline([('vect', CountVectorizer()),
               ('tfidf', TfidfTransformer()),
               ('clf', MultinomialNB()),
              ])
modelnb = nb.fit(X_train, y_train)

y_pred = nb.predict(X_test)

from sklearn.metrics import accuracy_score
print('accuracy %s' % accuracy_score(y_pred, y_test))

accuracy 0.8467505083214097


In [40]:
from sklearn.linear_model import LogisticRegression


logreg = Pipeline([('vect', CountVectorizer()),
                ('tfidf', TfidfTransformer()),
                ('clf', LogisticRegression(n_jobs=1, C=1e5)),
               ])
model_logreg = logreg.fit(X_train, y_train)



y_pred = logreg.predict(X_test)

print('accuracy %s' % accuracy_score(y_pred, y_test))

accuracy 0.8990134799307177


c:\users\nyach\appdata\local\programs\python\python38\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [41]:
#linear svm
from sklearn.linear_model import SGDClassifier

sgd = Pipeline([('vect', CountVectorizer()),
                ('tfidf', TfidfTransformer()),
                ('clf', SGDClassifier(loss='hinge', penalty='l2',alpha=1e-3, random_state=42, max_iter=5, tol=None)),
               ])
model_svm= sgd.fit(X_train, y_train)

y_pred = sgd.predict(X_test)

print('accuracy %s' % accuracy_score(y_pred, y_test))

accuracy 0.6168386173657655


In [42]:
from sklearn.ensemble import GradientBoostingClassifier
lr = Pipeline([('tfidf',TfidfVectorizer()),
               ('gbc',GradientBoostingClassifier())])
model_lr=lr.fit(X_train,y_train)
y_pred1 = lr.predict(X_test)

print(f"Accuracy is : {accuracy_score(y_pred1,y_test)}")

Accuracy is : 0.6747496046389035


Since from sklearn.linear_model, LogisticRegression has the maximum accuracy with 0.8990134799307177, using LogisticRegression to predict for Test_Data.csv

# Prediction

In [43]:
test=pd.read_csv("Test_Data.csv") #reading test data 

In [44]:
test

headline
0      area stand-up comedian questions the deal with...
1      dozens of glowing exit signs mercilessly taunt...
2      perfect response to heckler somewhere in prop ...
3                           gop prays for ossoff lossoff
4      trevor noah says the scary truth about trump's...
...                                                  ...
11061  house conservatives claim democrats have faile...
11062  area man having one of his little bursts of en...
11063   there is nothing libertarian about conservatives
11064  mike pompeo startled after seeing 'beware of h...
11065        how pets can help prevent suicide #nspw2017

[11066 rows x 1 columns]

In [47]:
test_list = list(test['headline']) #creating list of 'headline' column

In [48]:
prediction = [] #creating empty list to append predicted result

In [50]:
#to predict the 'is_sarcastic' that respective 'headline' may fall into
for i in test_list:
    headline=i
    pred = model_logreg.predict([repr(headline)])
    prediction.append(pred)

In [52]:
prediction= pd.DataFrame(prediction) #converting list into dataframe

In [53]:
prediction #prediction dataframe

0
0      1
1      1
2      1
3      0
4      0
...   ..
11061  0
11062  1
11063  1
11064  0
11065  0

[11066 rows x 1 columns]

In [54]:
prediction.columns=["prediction"] #renaming column 0 to "prediction"

In [56]:
prediction

prediction
0               1
1               1
2               1
3               0
4               0
...           ...
11061           0
11062           1
11063           1
11064           0
11065           0

[11066 rows x 1 columns]

In [58]:
 prediction.to_csv("Prediction.csv", index=False) #converting dataframe into csv file without index